README.md
## A minimal implementation of Transdirect API v4.

[![Build Status](https://travis-ci.org/sudiptpa/transdirect.svg?branch=master)](https://travis-ci.org/sudiptpa/transdirect)
[![StyleCI](https://styleci.io/repos/89677820/shield?branch=master)](https://styleci.io/repos/89677820)
[![Latest Stable Version](https://poser.pugx.org/sudiptpa/transdirect/v/stable?format=flat)](https://packagist.org/packages/sudiptpa/transdirect)
[![Total Downloads](https://poser.pugx.org/sudiptpa/transdirect/downloads?format=flat)](https://packagist.org/packages/sudiptpa/transdirect)
[![License](https://poser.pugx.org/sudiptpa/transdirect/license?format=flat)](https://packagist.org/packages/sudiptpa/transdirect)

Transdirect is Australia wide delivering solutions for you, this package covers a minimal PHP implementation of the Transdirect REST API. It contains only the endpoints documented at http://docs.transdirectapiv4.apiary.io.

### Installation

You can install the package via composer: [Composer](http://getcomposer.org/).

```
composer require sudiptpa/transdirect
```

And run composer to update your dependencies:

    $ curl -s http://getcomposer.org/installer | php
    $ php composer.phar update

### Usage

The first thing you need to do is get the API token from [Transdirect](https://www.transdirect.com.au/), you will need to specify the requesting domain to get valid API response data.

Here are a few examples on how you can use the package:

```php
  $client = new Sujip\Transdirect\Transdirect($apiKey);  
```
Also have a look in the [source code of `Sujip\Transdirect\Transdirect`](https://github.com/sudiptpa/transdirect/blob/master/src/Transdirect.php) to discover the methods you can use. You will need to visit [Official REST API documentation](http://docs.transdirectapiv4.apiary.io) for the parameters to specify with each end point.

If you wish to make a direct call to API end point for your own custom implementation, you can use the `$client->make('member', $parameters);` method.

```php
    $parameters = [
        'declared_value' => '1000.00',
        'referrer' => 'API',
        'requesting_site' => 'http://www.test.com.au',
        'tailgate_pickup' => true,
        'tailgate_delivery' => true,
        'items' => [
            ['weight' => '38.63',
                'height' => '0.25',
                'width' => '1.65',
                'length' => '3.32',
                'quantity' => 1,
                'description' => 'carton',
            ],
            [
                'weight' => '39.63',
                'height' => '1.25',
                'width' => '2.65',
                'length' => '4.32',
                'quantity' => 2,
                'description' => 'carton',
            ],
        ],
        'sender' => [
            'address' => '21 Kirksway Place',
            'company_name' => 'Test Company',
            'email' => 'sender@test.com',
            'name' => 'Sujip Thapa',
            'postcode' => '2000',
            'phone' => 123456789,
            'state' => '',
            'suburb' => 'SYDNEY',
            'type' => 'business',
            'country' => 'AU',
        ],
        'receiver' => [
            'address' => '216 Moggill Rd',
            'company_name' => 'Test Receiver',
            'email' => 'receiver@test.com',
            'name' => 'John Smith',
            'postcode' => '3000',
            'phone' => 123456789,
            'state' => '',
            'suburb' => 'MELBOURNE',
            'type' => 'business',
            'country' => 'AU',
        ],
    ];
    
  $response = $client->createBooking($parameters);
  $quotes = $response->getQuotes();
  $josn = $response->toJson();
  $bookingId = $response->getId();
  $statusCode = $response->getCode();
```

### Changelog

Please see [CHANGELOG](https://github.com/sudiptpa/transdirect/blob/master/CHANGELOG.md) for more information what has changed recently.

### Contributing

Contributions are **welcome** and will be fully **credited**.

Contributions can be made via a Pull Request on [Github](https://github.com/sudiptpa/paypal-ipn).



### Testing

If you would like to look at the real time response comming from live or mock server with transdirect REST API endpoints browse http://docs.transdirectapiv4.apiary.io



### Support

If you are having general issues with the package, feel free to drop me and email [sudiptpa@gmail.com](mailto:sudiptpa@gmail.com)

If you believe you have found a bug, please report it using the [GitHub issue tracker](https://github.com/sudiptpa/paypal-ipn/issues),
or better yet, fork the library and submit a pull request.


changes.md
 - vendor/bin/php-cs-fixer fix


tests/TransdirectTest.php
<?php

namespace Sujip\Transdirect\Test;

use GuzzleHttp\Client as GuzzleClient;
use PHPUnit\Framework\TestCase;
use Psr\Http\Message\ResponseInterface;
use Sujip\Transdirect\Transdirect;

class TransdirectTest extends TestCase
{
    /** @test */
    public function testInstanceOf()
    {
        $client = new Transdirect('cgfi7gagaf76');

        $this->assertInstanceOf(Transdirect::class, $client);
    }

    /**
     * @param $response
     * @param $endpoint
     * @param $parameters
     *
     * @return mixed
     */
    private function mockGuzzleRequest($response, $endpoint, $parameters)
    {
        $mockResponse = $this->getMockBuilder(ResponseInterface::class)
            ->getMock();
        $mockResponse->expects($this->once())
            ->method('getBody')
            ->willReturn($response);

        $mockGuzzle = $this->getMockBuilder(GuzzleClient::class)
            ->setMethods(['post'])
            ->getMock();
        $mockGuzzle->expects($this->once())
            ->method('post')
            ->with($endpoint, $parameters)
            ->willReturn($mockResponse);

        return $mockGuzzle;
    }
}


src/Response.php
<?php

namespace Sujip\Transdirect;

use GuzzleHttp\Message\Response as GuzzleHttpResponse;

/**
 * Class Response.
 */
class Response extends GuzzleHttpResponse
{
    /**
     * The guzzle http client response.
     *
     * @var \GuzzleHttp\Message\Response
     */
    protected $response;

    /**
     * Create a new response instance.
     *
     * @param GuzzleHttpResponse $response
     */
    public function __construct(GuzzleHttpResponse $response)
    {
        $this->response = $response;
    }

    public function getId()
    {
        $object = $this->toObject();

        return isset($object->id) ? $object->id : null;
    }

    /**
     * @return object
     */
    public function getItems()
    {
        $object = $this->toObject();

        if (!isset($object->items)) {
            return;
        }

        return $object->items;
    }

    /**
     * @return array
     */
    public function getQuotes()
    {
        $quotes = [];
        $object = $this->toObject();

        if (!isset($object->quotes)) {
            return;
        }

        if (is_object($object->quotes)) {
            foreach ($object->quotes as $key => $quote) {
                $formatted = sprintf(
                    '%s - %s [%s]',
                    ucwords(str_replace('_', ' ', $key)),
                    ucwords($quote->service),
                    $quote->transit_time
                );
                $quotes[] = [
                    'booking_id'         => $this->getId(),
                    'provider'           => $key,
                    'name_original'      => $this->parse($key),
                    'name_formatted'     => $this->parse($formatted),
                    'total'              => $quote->total,
                    'fee'                => $quote->fee,
                    'price_insurance_ex' => $quote->price_insurance_ex,
                    'insured_amount'     => (float) $quote->insured_amount,
                    'additional'         => 0,
                    'service'            => $quote->service,
                    'transit_time'       => $quote->transit_time,
                    'pickup_dates'       => $quote->pickup_dates,
                    'pickup_time'        => $quote->pickup_time,
                ];
            }
        }

        return $quotes;
    }

    public function toJson()
    {
        return (string) $this->response->getBody();
    }

    /**
     * @return mixed
     */
    public function toArray()
    {
        return json_decode($this->toJson(), true);
    }

    /**
     * @return mixed
     */
    public function toObject()
    {
        return json_decode($this->toJson());
    }

    /**
     * @return int
     */
    public function getCode()
    {
        return $this->response->getStatusCode();
    }

    /**
     * @param $string
     */
    public function parse($string)
    {
        $string = str_replace('_', ' ', $string);

        return str_replace('Tnt', 'TNT', $string);
    }
}


src/Transdirect.php
<?php

namespace Sujip\Transdirect;

use Sujip\Transdirect\Http\Request;

/**
 * Class Transdirect.
 */
class Transdirect extends Request
{
    /**
     * @param array $parameters
     *
     * @return \Sujip\Transdirect\Response
     */
    public function simpleQuotes(array $parameters)
    {
        return $this->make('quotes', $parameters);
    }

    /**
     * @param $parameters
     *
     * @return \Sujip\Transdirect\Response
     */
    public function createBooking($parameters)
    {
        return $this->make('bookings', $parameters);
    }

    /**
     * @param $since
     * @param null $sort
     *
     * @return \Sujip\Transdirect\Response
     */
    public function getBookings($since = null, $sort = null)
    {
        $uri = sprintf('bookings/?since=%s&sort=%s', $since, $sort);

        return $this->make($uri, [], 'get');
    }

    /**
     * @param $booking_id
     *
     * @return \Sujip\Transdirect\Response
     */
    public function getSingleBooking($booking_id)
    {
        $uri = sprintf('bookings/%s', $booking_id);

        return $this->make($uri, [], 'get');
    }

    /**
     * @param $booking_id
     * @param $parameters
     *
     * @return \Sujip\Transdirect\Response
     */
    public function updateBooking($booking_id, $parameters)
    {
        $uri = sprintf('bookings/%s', $booking_id);

        return $this->make($uri, $parameters, 'put');
    }

    /**
     * @param $booking_id
     *
     * @return \Sujip\Transdirect\Response
     */
    public function removeBooking($booking_id)
    {
        $uri = sprintf('bookings/%s', $booking_id);

        return $this->make($uri, [], 'delete');
    }

    /**
     * @param $booking_id
     * @param $parameters
     *
     * @return \Sujip\Transdirect\Response
     */
    public function confirmBooking($booking_id, $parameters)
    {
        $uri = sprintf('bookings/%s/confirm', $booking_id);

        return $this->make($uri, $parameters);
    }

    /**
     * @param $booking_id
     *
     * @return \Sujip\Transdirect\Response
     */
    public function trackBooking($booking_id)
    {
        $uri = sprintf('bookings/track/%s', $booking_id);

        return $this->make($uri, [], 'get');
    }

    /**
     * @param $booking_id
     *
     * @return \Sujip\Transdirect\Response
     */
    public function getBookingItems($booking_id)
    {
        $uri = sprintf('bookings/%s/items', $booking_id);

        return $this->make($uri, [], 'get');
    }

    /**
     * @param $booking_id
     * @param $parameters
     *
     * @return \Sujip\Transdirect\Response
     */
    public function addItemInBooking($booking_id, $parameters)
    {
        $uri = sprintf('bookings/%s/items', $booking_id);

        return $this->make($uri, $parameters);
    }

    /**
     * @param $page
     *
     * @return \Sujip\Transdirect\Response
     */
    public function getLocations($page = null)
    {
        $uri = 'locations';

        if (isset($page)) {
            $uri = sprintf($uri.'/page/%s', $page);
        }

        return $this->make($uri, [], 'get');
    }

    /**
     * @param $query
     *
     * @return \Sujip\Transdirect\Response
     */
    public function searchLocations($query = null)
    {
        $uri = sprintf('locations?q=%s', $query);

        return $this->make($uri, [], 'get');
    }

    /**
     * @param $query
     *
     * @return \Sujip\Transdirect\Response
     */
    public function getByPostcode($query)
    {
        $uri = sprintf('locations/postcode/%s', $query);

        return $this->make($uri, [], 'get');
    }

    /**
     * @param $parameters
     *
     * @return \Sujip\Transdirect\Response
     */
    public function createOrder($parameters)
    {
        return $this->make('orders', $parameters);
    }

    /**
     * @param $since
     * @param null $sort
     *
     * @return \Sujip\Transdirect\Response
     */
    public function getOrders($since = null, $sort = null)
    {
        $uri = sprintf('orders/?since=%s&sort=%s', $since, $sort);

        return $this->make($uri, [], 'get');
    }

    /**
     * @param $order_id
     *
     * @return \Sujip\Transdirect\Response
     */
    public function getOrder($order_id)
    {
        $uri = sprintf('orders/%s', $order_id);

        return $this->make($uri, [], 'get');
    }

    /**
     * @param $order_id
     * @param $parameters
     *
     * @return \Sujip\Transdirect\Response
     */
    public function updateOrder($order_id, $parameters)
    {
        $uri = sprintf('orders/%s', $order_id);

        return $this->make($uri, $parameters, 'put');
    }

    /**
     * @param $order_id
     *
     * @return \Sujip\Transdirect\Response
     */
    public function removeOrder($order_id)
    {
        $uri = sprintf('orders/%s', $order_id);

        return $this->make($uri, [], 'delete');
    }

    /**
     * @param $booking_id
     *
     * @return \Sujip\Transdirect\Response
     */
    public function getPdfLabel($booking_id)
    {
        $uri = sprintf('bookings/%s/label', $booking_id);

        return $this->make($uri, [], 'get');
    }

    /**
     * @param $booking_id
     *
     * @return \Sujip\Transdirect\Response
     */
    public function getInvoice($booking_id)
    {
        $uri = sprintf('bookings/%s/invoice', $booking_id);

        return $this->make($uri, [], 'get');
    }

    /**
     * @param $booking_id
     *
     * @return \Sujip\Transdirect\Response
     */
    public function getTntLabel($booking_id)
    {
        $uri = sprintf('bookings/%s/tntregeneralabel', $booking_id);

        return $this->make($uri, [], 'get');
    }

    /**
     * @return \Sujip\Transdirect\Response
     */
    public function getCouriers()
    {
        return $this->make('couriers', [], 'get');
    }

    /**
     * @return \Sujip\Transdirect\Response
     */
    public function getMember()
    {
        return $this->make('member', [], 'get');
    }
}


src/Http/Request.php
<?php

namespace Sujip\Transdirect\Http;

use GuzzleHttp\Client;
use GuzzleHttp\ClientInterface;
use Sujip\Transdirect\Endpoint;
use Sujip\Transdirect\Response;

/**
 * This is the client class.
 *
 * @author Sujip Thapa <support@sujipthapa.co>
 */
class Request
{
    use Endpoint;

    /**
     * The API token.
     *
     * @var string
     */
    protected $token;

    /**
     * The guzzle http client.
     *
     * @var \GuzzleHttp\ClientInterface
     */
    protected $client;

    /**
     * Create a new request instance.
     *
     * @param string             $token
     * @param \GuzzleHttp\Client $client
     */
    public function __construct($token, ClientInterface $client = null)
    {
        $this->token = $token;

        $this->client = $client ?: new Client();
    }

    /**
     * @param string $uri
     * @param array  $parameters
     * @param string $method
     *
     * @return \Sujip\Transdirect\Response
     */
    public function make($uri, array $parameters = [], $method = 'post')
    {
        $url = $this->getEndpoint($uri);

        $parameters = [
            'headers' => [
                'Content-Type' => 'application/json',
                'Api-Key' => $this->token,
            ],
            'body' => json_encode($parameters),
        ];

        try {
            $response = $this->client->{$method}($url, $parameters);
        } catch (ClientException $exception) {
            return $exception;
        }

        return new Response($response);
    }
}


live.php
<?php

use Sujip\Transdirect\Transdirect;

require __DIR__.'/vendor/autoload.php';

$transdirect = new Transdirect('1f5b8d2ff0820c7fbc24a8ba824e01a8');

$transdirect->sandbox();

$payload = [
    'declared_value'    => '1000.00',
    'referrer'          => 'API',
    'requesting_site'   => 'http://www.test.com.au',
    'tailgate_pickup'   => true,
    'tailgate_delivery' => true,
    'items'             => [
        ['weight'         => '38.63',
            'height'      => '0.25',
            'width'       => '1.65',
            'length'      => '3.32',
            'quantity'    => 1,
            'description' => 'carton',
        ],
        [
            'weight'      => '39.63',
            'height'      => '1.25',
            'width'       => '2.65',
            'length'      => '4.32',
            'quantity'    => 2,
            'description' => 'carton',
        ],
    ],
    'sender' => [
        'address'      => '21 Kirksway Place',
        'company_name' => 'Digital Thinker Pvt. Ltd.',
        'email'        => 'sudiptpa@gmail.com',
        'name'         => 'Sujip Thapa',
        'postcode'     => '2000',
        'phone'        => 123456789,
        'state'        => '',
        'suburb'       => 'SYDNEY',
        'type'         => 'business',
        'country'      => 'AU',
    ],
    'receiver' => [
        'address'      => '216 Moggill Rd',
        'company_name' => 'Digital Thinker Pvt. Ltd.',
        'email'        => 'support@digitalthinker.co',
        'name'         => 'John Smith',
        'postcode'     => '3000',
        'phone'        => 123456789,
        'state'        => '',
        'suburb'       => 'MELBOURNE',
        'type'         => 'business',
        'country'      => 'AU',
    ],
];

$confirm = [
    'courier'     => 'allied',
    'pickup-date' => '2015-08-27',
];

$item = [
    'weight'      => 11,
    'height'      => 11,
    'width'       => 11,
    'length'      => 11,
    'quantity'    => 5,
    'description' => 'carton',
];

$create_order = [
    'transdirect_order_id' => 12345,
    'order_id'             => '4444',
    'goods_summary'        => 'This is a test',
    'goods_dump'           => 'Another test',
    'imported_from'        => 'API',
    'purchased_time'       => '2015-06-01T16 =>06 =>52+1000',
    'sale_price'           => 10.2,
    'selected_courier'     => 'toll',
    'paid_time'            => '2015-06-01T16 =>06 =>52+1000',
    'buyer_name'           => 'John Doe',
    'buyer_email'          => 'john@testemail.com.au',
    'delivery'             => [
        'name'    => 'John Doe',
        'email'   => '',
        'phone'   => '123456789',
        'address' => '216 Moggill Rd',
    ],
];

$update_order = [
    'transdirect_order_id' => 12345,
    'order_id'             => '4444',
    'goods_summary'        => 'This is a test',
    'goods_dump'           => 'Another test',
    'imported_from'        => 'API',
    'purchased_time'       => '2015-06-01T16:06:52+1000',
    'sale_price'           => 10.2,
    'selected_courier'     => 'toll',
    'paid_time'            => '2015-06-01T16:06:52+1000',
    'buyer_name'           => 'John Doe',
    'buyer_email'          => 'john@testemail.com.au',
    'delivery'             => [
        'name'    => 'John Doe',
        'email'   => '',
        'phone'   => '123456789',
        'address' => '216 Moggill Rd',
    ],
];

$response = $transdirect->simpleQuotes($payload);

echo '<pre>';

print_r($response->toJson());
